### ISSUE TITLE
Migrate from setup.py to pyproject.toml

### ISSUE THREAD

**Issue by sigurdvaa** [alerta/alerta#1822]
# Migrate from setup.py to pyproject.toml

In the interest of the keeping the project up-to-date, we should consider migrating from the setup.py to the new standard with pyproject.toml.

## Python 3.12

In [Python 3.12](https://docs.python.org/3.12/whatsnew/3.12.html), the distutils module will be removed ([PEP 632](https://peps.python.org/pep-0632/)).
Setuptools has integrated a complete copy of distutils and does not depend on the standard library. Pip has silently been replacing distutils with setuptools for a long time.

## Setuptools

Starting with [PEP 621](https://peps.python.org/pep-0621/), the Python community selected pyproject.toml as a standard way of specifying project metadata. Setuptools has adopted this standard and will use the information contained in this file as an input in the build process.

Setuptools [discourage](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#setuppy-discouraged) the use of setup.py and that the majority of the command line interfaces are (or will be) deprecated.

If [compatibility with legacy](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html) builds or versions of tools that don’t support certain packaging standards (e.g. [PEP 517](https://peps.python.org/pep-0517/) or [PEP 660](https://peps.python.org/pep-0660/)), a simple setup.py script can be added to your project, while keeping the configuration in pyproject.toml

**Comment by satterly** [comment]
Perhaps I'm missing something here but we don't use `distutils` anywhere, right?  So we are unaffected by the removal of distutils from Python 3.12. And there are only a few places where `setup.py` is called directly, of which there are simple replacement commands.

So actually adopting pyproject.toml is a separate issue and should not be confused with distools or setup.py deprecation.

And considering how much work would be involved in migrating to toml files across all the python repos for Alerta I don't think we have the time or resources right now. We can target this for a future release but for now I would prefer to let others go through the pain first and we can learn from their experiences.

**Comment by sigurdvaa** [comment]
Sorry if the issue in unclear - it's about consideration. There's probably more reasons for or against it than I have mentioned.


No, we don't use `distutils`. But both `distutils` and `setuptools` use the concept of setup.py. And with `distutils` gone and `setuptools` discouraging the use of setup.py and PEP 621 making pyproject.toml the new standard, we might want to to migrate away from setup.py, as support for it is dwindling. We don't have to migrate all python repos at once, but perhaps one at a time is more appropriate.


The adoptation might belong in separate issues, I have no opinion here.


I do agree that time and resources are better spent on other issues for now, but I think a migration probably have to be done sooner or later.

**Comment by satterly** [comment]
For future reference...

<img width="764" alt="Screenshot 2023-03-18 at 09 11 51" src="https://user-images.githubusercontent.com/615057/226093909-63525cd9-c49e-4673-b449-f9e5cb10fc28.png">

https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html

**Comment by pantelis-karamolegkos** [comment]
Not sure is this relevant, but previously in `8.7.0` I built my custom image from the upstream `alerta-web` and installed my custom webhooks as follows

```Dockerfile
COPY webhooks-custom/ /app/webhooks-custom
RUN /venv/bin/pip install /app/webhooks-custom/my-webhook
```

This now fails with 

```
Running setup.py install for alerta-my-webhook did not run successfully.
  │ exit code: 1
  ╰─> [7 lines of output]
      running install
      /venv/lib/python3.8/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
        warnings.warn(
      running build
      running build_py
      creating build
      error: could not create 'build': Permission denied
      [end of output]
```

I worked around it by `USER root` (and then again `USER 1001`).

What caused this change? And why am I getting this deprecation warning since I am using `pip`

**Comment by satterly** [comment]
@pantelis-karamolegkos Yes, it is relevant. The change is because Python is deprecating the use of setup.py. We need to remove all references to `setup.py` before we move to Python 3.12 (which btw has not been released yet).

**Comment by satterly** [comment]
Good example here ... https://github.com/sphinx-doc/sphinx-intl/pull/86
